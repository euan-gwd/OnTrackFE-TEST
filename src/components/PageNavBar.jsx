import React from 'react';
import { connect } from 'react-redux';
import range from 'lodash/range';
import chunk from 'lodash/chunk';
import Pagination from 'react-bootstrap/Pagination';
import { getBooks } from '../store/actions';
import { useHistory } from 'react-router-dom';

export const PageNavBar = ({ dispatch, totalRecords, page, itemsPerPage }) => {
  let history = useHistory();
  const pageLimit = 20;
  const totalPages = Math.ceil(totalRecords / pageLimit);
  const pageRanges = range(1, totalPages);
  const pagesChunks = chunk(pageRanges, pageLimit);
  const [displayRanges] = pagesChunks.filter((item) => item.includes(page));
  const hasPages = displayRanges !== undefined ? displayRanges : false;

  const handlePageChange = (pg) => {
    const pageOptions = {
      page: pg,
      itemsPerPage: itemsPerPage
    };
    dispatch(getBooks(pageOptions));
    history.push(`/${pg}`);
  };

  const handleNext = () => {
    const pageOptions = {
      page: page + displayRanges.length,
      itemsPerPage: itemsPerPage
    };
    dispatch(getBooks(pageOptions));
    history.push(`/${pageOptions.page}`);
  };

  const handlePrev = () => {
    let newStartIndex = Number(page - pageLimit);
    if (newStartIndex < 0) {
      newStartIndex = 1;
    }
    const pageOptions = {
      page: newStartIndex,
      itemsPerPage: itemsPerPage
    };
    dispatch(getBooks(pageOptions));
    history.push(`/${pageOptions.page}`);
  };

  return (
    <>
      {hasPages ? (
        <Pagination>
          <Pagination.First onClick={() => handlePageChange(1)} disabled={page <= 1} />
          <Pagination.Prev disabled={page <= 1} onClick={() => handlePrev()} />
          {displayRanges.map((pg) => (
            <Pagination.Item key={pg} active={pg === page} onClick={() => handlePageChange(pg)} to={`/${pg}`}>
              {pg}
            </Pagination.Item>
          ))}
          <Pagination.Next onClick={() => handleNext()} disabled={page === totalPages - 1} />
          <Pagination.Last onClick={() => handlePageChange(totalPages - 1)} disabled={page === totalPages - 1} />
        </Pagination>
      ) : (
        <Pagination>
          <Pagination.First disabled />
          <Pagination.Prev disabled />
          <Pagination.Item active={1} onClick={() => handlePageChange(1)} to={`/${1}`}>
            Show All
          </Pagination.Item>
          <Pagination.Next disabled />
          <Pagination.Last disabled />
        </Pagination>
      )}
    </>
  );
};

const mapStateToProps = (state) => ({
  totalRecords: state.totalRecords,
  page: state.page,
  itemsPerPage: state.itemsPerPage
});

export default connect(mapStateToProps)(PageNavBar);
