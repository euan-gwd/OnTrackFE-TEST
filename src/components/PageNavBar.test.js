import React from 'react';
import { shallow } from 'enzyme';
import { PageNavBar } from './PageNavBar';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: 'localhost:3000/'
  }),
  useHistory: () => ({
    pathname: 'localhost:3000/'
  })
}));

describe('PageItemsFilter', () => {
  test('renders Pagination with records in array', () => {
    const props = {
      page: 1,
      itemsPerPage: 20,
      totalRecords: 100
    };

    const wrapper = shallow(<PageNavBar {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('renders Pagination with no records in array', () => {
    const props = {
      page: 1,
      itemsPerPage: 20,
      totalRecords: 1
    };

    const wrapper = shallow(<PageNavBar {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
