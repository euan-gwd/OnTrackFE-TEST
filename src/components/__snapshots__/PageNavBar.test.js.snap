// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageItemsFilter renders Pagination with no records in array 1`] = `
<Fragment>
  <ForwardRef>
    <First
      disabled={true}
    />
    <Prev
      disabled={true}
    />
    <PageItem
      active={1}
      activeLabel="(current)"
      disabled={false}
      onClick={[Function]}
      to="/1"
    >
      Show All
    </PageItem>
    <Next
      disabled={true}
    />
    <Last
      disabled={true}
    />
  </ForwardRef>
</Fragment>
`;

exports[`PageItemsFilter renders Pagination with records in array 1`] = `
<Fragment>
  <ForwardRef>
    <First
      disabled={true}
      onClick={[Function]}
    />
    <Prev
      disabled={true}
      onClick={[Function]}
    />
    <PageItem
      active={true}
      activeLabel="(current)"
      disabled={false}
      key="1"
      onClick={[Function]}
      to="/1"
    >
      1
    </PageItem>
    <PageItem
      active={false}
      activeLabel="(current)"
      disabled={false}
      key="2"
      onClick={[Function]}
      to="/2"
    >
      2
    </PageItem>
    <PageItem
      active={false}
      activeLabel="(current)"
      disabled={false}
      key="3"
      onClick={[Function]}
      to="/3"
    >
      3
    </PageItem>
    <PageItem
      active={false}
      activeLabel="(current)"
      disabled={false}
      key="4"
      onClick={[Function]}
      to="/4"
    >
      4
    </PageItem>
    <Next
      disabled={false}
      onClick={[Function]}
    />
    <Last
      disabled={false}
      onClick={[Function]}
    />
  </ForwardRef>
</Fragment>
`;

exports[`PageItemsFilter renders with no records in array 1`] = `
<Fragment>
  <ForwardRef>
    <First
      disabled={true}
    />
    <Prev
      disabled={true}
    />
    <PageItem
      active={1}
      activeLabel="(current)"
      disabled={false}
      onClick={[Function]}
      to="/1"
    >
      Show All
    </PageItem>
    <Next
      disabled={true}
    />
    <Last
      disabled={true}
    />
  </ForwardRef>
</Fragment>
`;

exports[`PageItemsFilter renders with records in array 1`] = `
<Fragment>
  <ForwardRef>
    <First
      disabled={true}
      onClick={[Function]}
    />
    <Prev
      disabled={true}
      onClick={[Function]}
    />
    <PageItem
      active={true}
      activeLabel="(current)"
      disabled={false}
      key="1"
      onClick={[Function]}
      to="/1"
    >
      1
    </PageItem>
    <PageItem
      active={false}
      activeLabel="(current)"
      disabled={false}
      key="2"
      onClick={[Function]}
      to="/2"
    >
      2
    </PageItem>
    <PageItem
      active={false}
      activeLabel="(current)"
      disabled={false}
      key="3"
      onClick={[Function]}
      to="/3"
    >
      3
    </PageItem>
    <PageItem
      active={false}
      activeLabel="(current)"
      disabled={false}
      key="4"
      onClick={[Function]}
      to="/4"
    >
      4
    </PageItem>
    <Next
      disabled={false}
      onClick={[Function]}
    />
    <Last
      disabled={false}
      onClick={[Function]}
    />
  </ForwardRef>
</Fragment>
`;
